File: answers/leetcode/Regular Expression Matching/Regular Expression Matching.cpp
//@result 445 / 445 test cases passed. Status: Accepted Runtime: 16 ms Submitted: 0 minutes ago You are here! Your runtime beats 55.16% of cpp submissions.

class Solution {
public:
    bool isMatch(string s, string p) {
        vector<vector<bool> > dp(s.length() + 1, vector<bool> (p.length() + 1, false));
        dp[0][0] = true;
        for (size_t i = 0; i <= s.length(); ++ i) {
            for (size_t j = 1; j <= p.length(); ++ j) {
                switch (p[j - 1]) {
                case '.':
                    dp[i][j] = i > 0 && dp[i - 1][j - 1];
                    break;
                case '*':
                    if (j > 1) {
                        dp[i][j] = dp[i][j - 2] || (i > 0 && dp[i - 1][j] && (s[i - 1] == p[j - 2] || '.' == p[j - 2]));
                    }
                    break;
                default:
                    dp[i][j] = i > 0 && dp[i - 1][j - 1] && s[i - 1] == p[j - 1];
                    break;
                }
            }
        }
        return dp.back().back();
    }
};

File: answers/leetcode/Regular Expression Matching/Regular Expression Matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        boolean[][] dp = new boolean[s.length() + 1][p.length() + 1];
        dp[0][0] = true;
        for (int i = 1; i < p.length(); i++) {
            if (p.charAt(i) == '*') {
                dp[0][i + 1] = dp[0][i - 1];
            }
        }
        for (int i = 0; i < s.length(); i++) {
            for (int j = 0; j < p.length(); j++) {
                if (p.charAt(j) == '.') {
                    dp[i + 1][j + 1] = dp[i][j];
                } else if (p.charAt(j) == '*' && j > 0) {
                    if (s.charAt(i) == p.charAt(j - 1) || p.charAt(j - 1) == '.') {
                        dp[i + 1][j + 1] = dp[i][j + 1] || dp[i + 1][j] || dp[i + 1][j - 1];
                    } else {
                        dp[i + 1][j + 1] = dp[i + 1][j - 1];
                    }
                } else {
                    dp[i + 1][j + 1] = s.charAt(i) == p.charAt(j) ? dp[i][j] : false;
                }
            }
        }
        return dp[s.length()][p.length()];
    }
}